### PR TITLE
Delete a shopcart API

### PR DESCRIPTION
**Pull request by nijya**
1. Update logic part in `route.py`.
2. Create tests in `test_routes.py` and `test_models.py`.
3. Resolve some existing linter errors

test-coverage

linter-score

### CODE REVIEW

File: service/models.py [hunk 1]
```diff
@@ -157,7 +157,7 @@ def find_id(cls):
         for res in result:
             ids.append(res.id)
         return len(ids)
-
+
```
Comment (wl2484) [inline service/models.py:160]: I remember a whitespace here would trigger one of the linter error 😆

File: service/routes.py [hunk 2]
```diff
@@ -115,6 +115,26 @@ def get_shopcarts(shopcart_id):
...
+def delete_shopcart(shopcart_id):
+    """Deletes a shopcart
+        Args:
+            user_id (str): the user_id of the shopcart to delete
+        Returns:
+            str: always returns an empty string
+    """
+    app.logger.info("Start deleting shopcart %s...", shopcart_id)
+    shopcart = Shopcart.get_by_id(shopcart_id)
+    if shopcart:
```
Comment (wl2484) [inline service/routes.py:132]: we need to return 404 Not Found when shopcart_id does not exist

File: tests/test_routes.py [hunk 3]
```diff
@@ -291,3 +293,45 @@ def test_add_shopcart_items_data_validation_error(self, mock_shopcart_update):
...
+        res = self.client.delete(f"{BASE_URL}/{shopcart_id}")
+        self.assertEqual(res.status_code, status.HTTP_204_NO_CONTENT)
+        # Make sure shopcart is deleted
+        res = self.client.get(f"{BASE_URL}/{shopcart_id}")
+        self.assertEqual(res.status_code, status.HTTP_404_NOT_FOUND)
+
+    def test_delete_invalid_shopcart(self):
+        """It Should Delete an non-existing shopcart"""
+        res = self.client.delete(f"{BASE_URL}/{NONEXIST_SHOPCART_ID}")
+        self.assertEqual(res.status_code, status.HTTP_204_NO_CONTENT)
```
Comment (wl2484) [inline tests/test_routes.py:337]: status code should be **404 Not Found** if it's a non-existent `shopcart_id`

File: service/routes.py (hunk 2 above)
Comment (wl2484) [reply service/routes.py:132]: also, you would need a corresponding test case like [what @momo100110 Kangle did](https://github.com/CSCI-GA-2820-SU23-001/shopcarts/blob/master/tests/test_routes.py#L556) for the Delete Shopcart Item API to remove line  133-134 from lines missing test cases.

**Review by wl2484 — CHANGES_REQUESTED** [review]

**Comment by wl2484** [comment]
besides, looks like there's conflicts with the master branch, please help to resolve 🙇‍♀️

File: service/routes.py [hunk 4]
```diff
@@ -115,6 +115,26 @@ def get_shopcarts(shopcart_id):
...
+        shopcart.delete()
+        app.logger.info("Shopcart deleted with id= %s ", shopcart_id)
+    return make_response("", status.HTTP_204_NO_CONTENT)
```
Comment (wl2484) [inline service/routes.py:135]: In addition, please help to update the [Delete Shopcart API doc](https://github.com/CSCI-GA-2820-SU23-001/shopcarts/blob/master/README.md?plain=1#L284), thanks! 🙇‍♀️

File: service/routes.py (hunk 2 above)
Comment (nijya) [reply service/routes.py:132]: 🤔 According to lecture slide 03-Building-RESTful-Services page43, does delete always return 204_NO_CONTENT?

Comment (wl2484) [reply service/routes.py:132]: I double checked the slide, and you're right. Sorry to confuse you.

Comment (wl2484) [reply service/routes.py:132]: This reminds me that the Delete Shopcart API doc needs to be revised accordingly, please help to update the doc, thanks!

Comment (momo100110) [reply service/routes.py:132]: So I need to modify here in Delete A Shopcart Item API as well, right? @nijya @wl2484

Comment (nijya) [reply service/routes.py:132]: > 

Yes

File: tests/test_routes.py [hunk 5]
```diff
@@ -556,20 +558,45 @@ def test_delete_item_not_found(self):
     def test_delete_shopcart_item_with_none_shopcart(self):
```
Comment (wl2484) [inline tests/test_routes.py:558]: Found most part of this test case was removed accidentally when resolving the conflict.

File: service/models.py (hunk 1 above)
Comment (wl2484) [reply service/models.py]: I added a commit to correct this and also other linter errors.

File: tests/test_routes.py (hunk 5 above)
Comment (wl2484) [reply tests/test_routes.py:558]: Hence I added a commit to bring it back lol

**Review by wl2484 — APPROVED** [review]
LGTM

**Pull request merged**